### ISSUE TITLE
TrackMe QOS - Time period for SLA not consistent with total duration

### ISSUE THREAD

**Issue by s2835648** [guilhemmarchand/trackme#245]
When accessing the TrackMe - Quality Of Service Auditing Dashboard, the time period for SLA calculation is not consistent with the overall total duration.

For example, when I choose the Last 30 days from the time picker, the time spent in green state + time in non-green state, totals 60+ days. From my understanding, the total duration should never exceed the selected timeframe. This behavior is also seen when I choose any timeframe from the picker.

![image](https://user-images.githubusercontent.com/7518441/107922357-2c964d00-6f70-11eb-8c8d-829fd187f27c.png)

Hopefully this make sense. Not sure if this is expected behavior or a bug :)

**Comment by guilhemmarchand** [comment]
Thanks @s2835648 for reporting it and looking at this.
You are actually right, I knew I shouldn't have trusted the stats range function!

**I have been quickly reviewing it, something like:**

```
index="trackme_summary" source="current_state_tracking:data_source" object_category="data_source" object="network:pan:traffic" object="*" 
| fields _time object_category, object, current_state 
| stats first(current_state) as current_state by _time, object_category, object 
| streamstats last(_time) as "prev_time", last(current_state) as prev_state current=f by object_category, object
| eval range_duration=_time-prev_time
| eval green_time=case(current_state="green" OR current_state="blue", range_duration), not_green_time=case(current_state!="green" AND current_state!="blue", range_duration) 
| stats sum(range_duration) as range_duration, sum(green_time) as green_time, sum(not_green_time) as not_green_time by object_category, object 
| foreach green_time not_green_time 
    [ eval <<FIELD>> = if(isnum('<<FIELD>>'), '<<FIELD>>', 0) ] 
| eval percent_sla=round(green_time/range_duration*100, 2) 
| foreach range_duration green_time not_green_time 
    [ eval <<FIELD>> = tostring('<<FIELD>>', "duration") ] 
| lookup trackme_data_source_monitoring data_name as object, object_category OUTPUTNEW priority, data_monitored_state as monitored_state 
| lookup trackme_host_monitoring data_host as object, object_category OUTPUTNEW priority, data_monitored_state as monitored_state 
| lookup trackme_metric_host_monitoring metric_host as object, object_category OUTPUTNEW priority, metric_monitored_state as monitored_state 
| search priority="*" 
| fields object, object_category, priority, monitored_state, range_duration, green_time, not_green_time, percent_sla 
| rename range_duration as "total duration ([D+]HH:MM:SS)", green_time as "duration spent in green state ([D+]HH:MM:SS)", not_green_time as "duration spent in non green state ([D+]HH:MM:SS)"
```

Works as it expected, it is basically the expansion from the macro with a fix to use streamstats rather than a stats range approach.

I will carefully review this.

Guilhem

**Comment by guilhemmarchand** [comment]
@s2835648 

This will be fixed in upcoming version 1.2.35, currently in testing branch:

https://github.com/guilhemmarchand/trackme/tree/testing

Using a streamstats based approach produces the expected consistent results.

Thanks

**Comment by guilhemmarchand** [comment]
Version 1.2.35 is now live.
